code/modules/projectiles/guns/projectile/launchers.dm

```diff
@@ -83,3 +83,752 @@
+	var/yd = loc.y // Y location of next tile to move to.
+	transform = turn(transform, Angle)
+	
+	while(loc)
+		xd += sin(Angle)*spd_multi
+		yd += cos(Angle)*spd_multi
+		if(round(xd, 1) <= 0 || round(xd,1) >= world.maxx)
+			on_range()
+			break
+		if(round(yd, 1) <= 0 || round(yd, 1) >= world.maxy)
+			on_range()
+			break
+		new /obj/effect/temp_visual/smoke_puff(loc)
+		var/turf/next_t = locate(round(xd, 1), round(yd, 1), z)
+		if(prehit(next_t))
+			loc = next_t
+			Range()
+		else
+			break
+		sleep(max(1, speed))
+		
+/obj/item/projectile/a66mm/prehit(turf/target as turf)
+	// Check for anything that stops the rocket.
+	for(var/i in rocket_stop)
+		if(istype(target, i))
+			on_hit(target)
+			qdel(src)
+			return FALSE
+		for(var/atom/j in target.contents)
+			if(istype(j, i) && j.density)
+				on_hit(target)
+				qdel(src)
+				return FALSE
+
+	// Anything that didn't stop the rocket, destroy it.
+	var/fx_hit_flag = FALSE
+	for(var/obj/j in target.contents)
+		if(src == j)
+			continue
+		var/skip_obj = FALSE
+		for(var/i in rocket_stop)
+			if(istype(j, i))
+				skip_obj = TRUE
+				break
+		if(skip_obj)
+			continue
+		if(istype(j, /obj/machinery/door))
+			j.take_damage(j.obj_integrity)
+			fx_hit_flag = TRUE
+		if(istype(j, /obj/machinery/door/airlock/multi_tile))
+			qdel(j)
+			fx_hit_flag = TRUE
+		else if(j.density)
+			j.take_damage(j.obj_integrity)
+			fx_hit_flag = TRUE
+			
+	if(fx_hit_flag)
+		sfx_hit(50, 33)
+		ffx_hit(target)
+
+	// Destroy structures or anything that was spawned from removal.
+	for(var/obj/j in target.contents)
+		var/skip_obj = FALSE
+		for(var/i in rocket_stop)
+			if(istype(j, i))
+				skip_obj = TRUE
+				break
+		if(skip_obj)
+			continue
+		if(istype(j, /obj/machinery/constructable_frame))
+			qdel(j)
+		else if(istype(j, /obj/structure/grille/broken))
+			continue
+		else if(istype(j, /obj/structure/table_frame))
+			continue
+		else if(istype(j, /obj/structure))
+			if(j.density)
+				qdel(j)
+
+	return TRUE
+	
+/obj/item/projectile/a66mm/on_hit(turf/target, blocked = 0)
+	if(locate(/obj/mecha) in target.contents)
+		for(var/obj/mecha/M in target.contents)
+			if(!is_fellow_antag(M) && M.density)
+				M.take_damage(damage, BRUTE)
+				break
+	else if(locate(/mob/living) in target.contents)
+		for(var/mob/living/L in target.contents)
+			if(!is_fellow_antag(L) && L.density)
+				L.apply_damage(damage, BRUTE)
+				L.check_projectile_dismemberment(src, "chest")
+				break
+	else if(istype(target, /turf/simulated/wall/r_wall))
+		var/turf/simulated/wall/r_wall/W = target
+		W.add_dent(WALL_DENT_HIT)
+		target = get_turf(src)
+	else if(istype(target, /turf/simulated/wall))
+		var/turf/simulated/wall/W = target
+		W.break_wall()
+		target = get_turf(src)
+	else if(locate(/obj/structure/girder) in target.contents)
+		var/obj/structure/girder/G = locate(/obj/structure/girder) in target.contents
+		G.deconstruct(FALSE)
+		target = get_turf(src)
+	else
+		target = get_turf(src)
```

Comment:
Can extract this line from the else and the 3 ifs above here. Less dupe code the better